### PR TITLE
T181234 Scroll to fieldsets

### PR DESCRIPTION
**Pull request by gbirke**
When clicking on a radio button, the form scrolls to the related
section.

* Moved all scrolling into a library function. When different events cause scroll animations, the animations are now interrupted.
* Refactored "SmoothScroll" function out of main.js

### CODE REVIEW

File: skins/cat17/src/app/lib/animated_scroller.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		scrollToTarget: function( evt ) {
+			evt.preventDefault();
+			if ( this.linkIsInsideCompletedSummaryOnSmallScreen( evt.currentTarget ) || this.linkIsOnDifferentPage( evt.currentTarget ) ) {
+				return;
+			}
+			var target = $( evt.currentTarget.hash );
+			target = target.length ? target : $( '[name=' + evt.currentTarget.hash.slice(1) + ']' );
+			// Does a scroll target exist?
+			if ( target.length ) {
+				// Only prevent default if animation is actually gonna happen
```
Comment (wiese) [inline skins/cat17/src/app/lib/animated_scroller.js]: We already prevented default in line 45. Which line is superfluous?

File: skins/cat17/src/scripts/donationForm.js [hunk 2]
```diff
@@ -338,4 +337,12 @@ $( function () {
     $introDefault.addClass('hidden');
   }
 
+	// Non-state-changing event behavior
+
+	var scroller = WMDE.AnimatedScroller.createAnimatedScroller( $( '.wrap-header, .state-bar' ) );
+	WMDE.AnimatedScroller.addScrollToLinkAnchors( $('a[href*="#"]'), scroller);
+	WMDE.AnimatedScroller.scrollOnSuboptionChange( $( 'input[name="addressType"]' ), $('#type-donor'), scroller );
+	WMDE.AnimatedScroller.scrollOnSuboptionChange( $( 'input[name="paymentType"]' ), $( '#donation-payment' ), scroller );
+
```
Comment (wiese) [inline skins/cat17/src/scripts/donationForm.js]: We briefly talked about scrolling to the interval texts as well. Is that off the table?

File: skins/cat17/src/scripts/main.js [hunk 3]
```diff
@@ -1,7 +1,6 @@
 (function ($) {
 
     var init = function () {
-        smoothScroll();
```
Comment (wiese) [inline skins/cat17/src/scripts/main.js]: Please add a note in membershipForm, lest we forget.

File: skins/cat17/src/app/lib/animated_scroller.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		}
+	};
+
+module.exports ={
+	createAnimatedScroller: function ( fixedHeaderElements ) {
+		return objectAssign( Object.create( AnimatedScroller ), { fixedHeaderElements: fixedHeaderElements } );
+	},
+	scrollOnSuboptionChange: function( $suboptionInput, $suboptionContainer, scroller ) {
+		$suboptionInput.on( 'change', function ( evt ) {
+			var wrapper = $suboptionContainer.find( '.wrap-field.' + evt.target.value );
```
Comment (wiese) [inline skins/cat17/src/app/lib/animated_scroller.js]: That concat looks briddle; I'm surprised it works at all. Goal reminds me of [this](https://github.com/wmde/FundraisingFrontend/blob/master/skins/cat17/src/app/lib/view_handler/display_field_suboptions.js#L15). Can we reasonably extract that match, `value` => `container`/ `info` / ... ?

File: skins/cat17/src/app/lib/animated_scroller.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+var objectAssign = require( 'object-assign' ),
+	AnimatedScroller = {
+		fixedHeaderElements: null,
+		scrollTo: function( $element ) {
+			var offset = 0,
+				matchedElemPadding = $element.css( 'padding-top' ).match( /^(\d+)px$/ );
+			this.fixedHeaderElements.each( function ( idx, elm ) {
```
Comment (wiese) [inline skins/cat17/src/app/lib/animated_scroller.js]: With `_.each` it [becomes testable](https://github.com/wmde/FundraisingFrontend/blob/master/skins/cat17/src/app/tests/test_form_data_extractor.js#L18).

File: skins/cat17/src/scripts/donationForm.js (hunk 2 above)
Comment (gbirke) [reply skins/cat17/src/scripts/donationForm.js]: I just forgot and added it in 3ae28a572668a7d766b00a4a663448f9a4ce770b.

However, it feels a bit weird and I'd like @jdittrich to have a look

File: skins/cat17/src/scripts/main.js (hunk 3 above)
Comment (gbirke) [reply skins/cat17/src/scripts/main.js]: Done in e72fed42f44f9fd15d752ca09fcaa6be2b1073c9

**Pull request merged**